### PR TITLE
security: don't hash empty password

### PR DESCRIPTION
**Pull request by RaduBerinde**
Avoid hashing empty password (which happens when creating users with no
password). This makes the sql tests faster. See #11746.

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/11781)

### CODE REVIEW

**Comment by bdarnell** [comment]
I think I'd rather push this up a level (although I haven't reacquainted myself with the relevant code): don't call HashPassword if there is no password, instead of calling HashPassword on an empty string.

**Comment by RaduBerinde** [comment]
Good call. Done.

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11781#-:-KY-KC21i0h8NuRcUlPq:bombmv2)*

**Comment by asubiotto** [comment]
I see your point but I'm not sure how much good that will do and will lead to some repetitive code. Making it explicit that a valid hash is only returned for non-empty password might be enough.

**Comment by bdarnell** [comment]
LGTM. Just make sure that we have test coverage to ensure that if a user doesn't have a password, we don't allow login with an empty password.

**Comment by asubiotto** [comment]
LGTM. That test already exists at the bottom of `pkg/sql/pgwire_test.go`

**Comment by RaduBerinde** [comment]
One more thing, regarding the behavior of `PromptForPasswordAndHash` - if a user enters an empty password, this is accepted but it means that they will no longer be able to log in by password. Should we error out on empty passwords instead?

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11781#-:-KY-Mk78p21T9uwGLAJM:b-b78ar)*

**Comment by asubiotto** [comment]
Good catch. We should definitely error out with something like "empty passwords are not permitted" in PromptForPassword.

**Comment by RaduBerinde** [comment]
Done.

The grammar disallows that (I verified manually). Added a TestParseError case.

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11781#-:-KY-QfaTRTaSRilCUnXC:b-zzx4f)*

**Comment by RaduBerinde** [comment]
Ah sorry it would be `WITH PASSWORD ''`, i'll fix that.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11781#-:-KY-RWjIAgMOiHIAj6Zk:bzhuqif)*

**Comment by RaduBerinde** [comment]
Ah, `WITH PASSWORD ''` actually makes the password `''`. So we only care about what you said, the statement ending in `WITH PASSWORD`, and that is covered.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11781#-:-KY-T1G7SnaMoYszVDOh:b-3z3a7k)*

File: pkg/security/password.go [hunk 1]
```diff
@@ -52,6 +52,9 @@ func PromptForPassword() (string, error) {
 	if err != nil {
 		return "", err
 	}
+	if len(one) == 0 {
+		return "", errors.New("empty passwords not permitted")
```
Comment (tamird) [inline pkg/security/password.go]: should this be tested?

**Comment by RaduBerinde** [comment]
*[pkg/security/password.go, line 56 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11781#-KY-HJXkUZ1vxQNlKF6c-r1-56:-KY-V-fxXaKa_16g6X_P:bb6x76c) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d906404582b3ea6fbd7ad681f908f90ca9f2ed6/pkg/security/password.go#L56)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

should this be tested?
</blockquote></details>

How?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11781)*

**Comment by tamird** [comment]
Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/security/password.go, line 56 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11781#-KY-HJXkUZ1vxQNlKF6c-r1-56:-KY-V859Fd3so2igNtYj:bloowel) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d906404582b3ea6fbd7ad681f908f90ca9f2ed6/pkg/security/password.go#L56)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

How?
</blockquote></details>

docker CLI test

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11781)*

**Comment by RaduBerinde** [comment]
Added commit to disallow empty `WITH PASSWORD` values.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11781#-:-KY-Vm-M6qLfXkQsXTU6:b-o32k00)*

**Comment by RaduBerinde** [comment]
*[pkg/security/password.go, line 56 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11781#-KY-HJXkUZ1vxQNlKF6c-r1-56:-KY-W77voH5w7kSFPehp:b-a7d4xi) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d906404582b3ea6fbd7ad681f908f90ca9f2ed6/pkg/security/password.go#L56)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

docker CLI test
</blockquote></details>

@asubiotto - can you do that in a follow-up change? I already spent way more time on this than I was planning.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11781)*

**Comment by asubiotto** [comment]
Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/security/password.go, line 56 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11781#-KY-HJXkUZ1vxQNlKF6c-r1-56:-KY-WxhSxxuJJVTJ_o-g:b-3m0sdf) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d906404582b3ea6fbd7ad681f908f90ca9f2ed6/pkg/security/password.go#L56)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

@asubiotto - can you do that in a follow-up change? I already spent way more time on this than I was planning.
</blockquote></details>

Sure. Thanks for what you've done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11781)*

**Comment by asubiotto** [comment]
Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/security/password.go, line 56 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11781#-KY-HJXkUZ1vxQNlKF6c-r1-56:-KY-hFpUd5puGp1Agq5C:bpxojod) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d906404582b3ea6fbd7ad681f908f90ca9f2ed6/pkg/security/password.go#L56)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Sure. Thanks for what you've done.
</blockquote></details>

Just added.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11781)*

**Comment by tamird** [comment]
:lgtm: modulo the inconsistent error string.

---

Reviewed 2 of 4 files at r1, 1 of 2 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/create.go, line 302 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11781#-KY-sLeMP1kQtb24ijAi:-KY-sLeMP1kQtb24ijAj:b7al8ex) ([raw file](https://github.com/cockroachdb/cockroach/blob/186300448b6e458218b10696461a44a7fa2d1a2e/pkg/sql/create.go#L302)):*
> ```Go
> 		resolvedPassword = string(*password.(*parser.DString))
> 		if resolvedPassword == "" {
> 			return nil, errors.New("empty passwords are not permitted")
> ```

wow, it is super confusing that this string is ever so slightly different than the one in security/password.go.

can you extract these two into a common string?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11781#-:-KY-sVAao1o9BJUpV-ES:b-reoyxm)*

**Comment by RaduBerinde** [comment]
Review status: 3 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/sql/create.go, line 302 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11781#-KY-sLeMP1kQtb24ijAi:-KY08Vzac_4z3XZEmZQ9:boj1qgt) ([raw file](https://github.com/cockroachdb/cockroach/blob/186300448b6e458218b10696461a44a7fa2d1a2e/pkg/sql/create.go#L302)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

wow, it is super confusing that this string is ever so slightly different than the one in security/password.go.

can you extract these two into a common string?
</blockquote></details>

Good catch. Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11781)*

**Comment by tamird** [comment]
Reviewed 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/security/password.go, line 36 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/11781#-KY09_tmGtOzwr_9rvXc:-KY09_tmGtOzwr_9rvXd:bb6fwnr) ([raw file](https://github.com/cockroachdb/cockroach/blob/8eda0a52ba67fa637cbdec7c41bbed898c07ddab/pkg/security/password.go#L36)):*
> ```Go
> const bcryptCost = bcrypt.DefaultCost
> 
> var ErrEmptyPassword = errors.Errorf("empty passwords are not permitted")
> ```

.New

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11781)*

**Comment by RaduBerinde** [comment]
Review status: 5 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/security/password.go, line 36 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/11781#-KY09_tmGtOzwr_9rvXc:-KY09uOinjbLs5fhPCmD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8eda0a52ba67fa637cbdec7c41bbed898c07ddab/pkg/security/password.go#L36)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

.New
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11781)*

**Comment by tamird** [comment]
:shipit: 

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11781#-:-KY0A5Wp2ooSW83H0sTg:bizlbj6)*

**Pull request merged**